src/Workspaces/Core/Portable/FindSymbols/FindReferences/DependentTypeFinder.cs

```diff
@@ -134,151 +144,19 @@ internal static partial class DependentTypeFinder
         }
 
         /// <summary>
-        /// Used for implementing the Inherited-By relation for progression.
-        /// </summary>
-        public static Task<ImmutableArray<INamedTypeSymbol>> FindImmediatelyDerivedClassesAsync(
-            INamedTypeSymbol type,
-            Solution solution,
-            CancellationToken cancellationToken)
-        {
-            return FindTypesFromCacheOrComputeAsync(
-                type, solution, projects: null,
-                cache: s_typeToImmediatelyDerivedClassesMap,
-                findAsync: c => FindDerivedClassesAsync(type, solution, projects: null,
-                    transitive: false, cancellationToken: c),
-                cancellationToken: cancellationToken);
-        }
-
-        /// <summary>
-        /// This is an internal implementation of <see cref="SymbolFinder.FindDerivedClassesAsync(INamedTypeSymbol, Solution, IImmutableSet{Project}, CancellationToken)"/>, which is a publically callable method.
+        /// Walks down a <paramref name="type"/>'s inheritance tree looking for more <see cref="INamedTypeSymbol"/>'s
+        /// that match the provided <paramref name="typeMatches"/> predicate.
         /// </summary>
-        public static Task<ImmutableArray<INamedTypeSymbol>> FindTransitivelyDerivedClassesAsync(
-            INamedTypeSymbol type,
-            Solution solution,
-            IImmutableSet<Project> projects,
-            CancellationToken cancellationToken)
-        {
-            return FindTypesFromCacheOrComputeAsync(
-                type, solution, projects, s_typeToTransitivelyDerivedClassesMap,
-                c => FindDerivedClassesAsync(type, solution, projects, transitive: true, c),
-                cancellationToken);
-        }
-
-        private static Task<ImmutableArray<INamedTypeSymbol>> FindDerivedClassesAsync(
+        /// <param name="shouldContinueSearching">Called when a new match is found to check if that type's inheritance
+        /// tree should also be walked down.  Can be used to stop the search early if a type could have no types that
+        /// inherit from it that would match this search.</param>
+        /// <param name="transitive">If this search after finding the direct inherited types that match the provided
+        /// predicate, or if the search should continue recursively using those types as the starting point.</param>
+        private static async Task<ImmutableArray<INamedTypeSymbol>> DescendInheritanceTreeAsync(
             INamedTypeSymbol type,
             Solution solution,
             IImmutableSet<Project> projects,
-            bool transitive,
-            CancellationToken cancellationToken)
```

Comment:
renamed this for clarity.  'Find' was a little vague.  i think 'DescendInheritanceTree' gives more of an intuition about what is going on here.